12-offer-encoding.md

```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
+    - if `offer_paths` is not set or does not contain at least one `blinded_path`:
       - MUST NOT respond to the offer.
     - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the `node_id`.
-  - if `signature` is present, but is not a valid signature using `node_id` as described in [Signature Calculation](#signature-calculation):
-    - MUST NOT respond to the offer.
-  - if it uses `amount` to provide the user with a cost estimate:
+  - if it uses `offer_amount` to provide the user with a cost estimate:
     - MUST warn user if amount of actual invoice differs significantly
         from that expectation.
   - SHOULD not respond to an offer if the current time is after
-    `absolute_expiry`.
+    `offer_absolute_expiry`.
   - FIXME: more!
 
 ## Rationale
 
-A signature is optional because it makes for a longer string (potentially
+The entire offer is reflected in the invoice_request, both for
+completeness (so all information will be returned in the invoice), and
+so that the offer node can be stateless.  This makes `offer_metadata`
+particularly useful, since it can contain an authentication cookie to
+validate the other fields.
+
+A signature is unnecessary, and makes for a longer string (potentially
 limiting QR code use on low-end cameras); if the offer has an error, no
 invoice will be given (or, for `send_invoice` offers, accepted) since
 the `offer_id` already covers all the non-signature fields.
 
-The `node_id` is redundant if a `blinded_path` is provided: a public
-node can always arrange that the final `node_id` in the path is
-unblinded.
+If `node_id` is anonymous, it might as well be selected from `blinded_path`
+(the final node will know how to tweak its real node id to provide signatures for this blinded id).
 
 # Invoice Requests
 
 Invoice Requests are a request for an invoice; the human-readable prefix for
-invoices is `lnr`.
+invoices is `lnr`.  It mirrors all the fields from the offer, except
+`offer_send_invoice` which cannot cause invoice_requests.
+
 
 ## TLV Fields for `invoice_request`
 
 1. `tlv_stream`: `invoice_request`
 2. types:
-    1. type: 3 (`chain`)
+    1. type: 0 (`invoice_request_payer_info`)
     2. data:
-        * [`chain_hash`:`chain`]
-    1. type: 4 (`offer_id`)
+        * [`...*byte`:`blob`]
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
```

Comment:
I'm wondering if we need to mirror this particular field in `invoice_request`s and `invoice`s? It seems like the biggest potential bandwidth consumer in the mirroring, and the other mirrored fields seem more relevant to determining whether an incoming `inv_req` is valid 🤔

Reply:
Sure, but simple rules are good.  I mean, we could have a non-mirrored range, too.

I suppose a really dumb implementation could reuse these blinded paths for the invoice (they have to decrypt and re-encrypt them with more fields, which I *think* is possible if you store the blinding secret (or enough to derive it) in offer_metadata?).

Reply:
Not sure I followed how these could be reused for the invoice, but I'm good with mirroring all fields for simplicity :)